### PR TITLE
Add a C++ usage guidelines page

### PR DESCRIPTION
**Pull request by Calinou**
This page describes C++ features that are used and allowed in Godot's codebase. It's based on community consensus over the last months, as Godot recently moved to C++17 in the `master` branch.

I've probably missed many things, so please leave comments. That said, the C++ style we expect is now (mostly) set in stone, so most disallowed features written here are not negotiable (such as ``auto``). Please keep the discussion on-topic :slightly_smiling_face:

I know @hpvb intended to write this page at first, but I decided to do it to help speed up the process.

This closes https://github.com/godotengine/godot/issues/9694.

### CODE REVIEW

File: community/contributing/cpp_usage_guidelines.rst
Comment (Calinou) [inline community/contributing/cpp_usage_guidelines.rst:89]: I wrote the part about lambdas this way because we have a few instances of lambdas in the current editor code. These could be rewritten into macros, but it's not really convenient.

File: community/contributing/cpp_usage_guidelines.rst
Comment (asmaloney) [inline community/contributing/cpp_usage_guidelines.rst:48]: "doesn't" -> "don't"

Comment (Calinou) [reply community/contributing/cpp_usage_guidelines.rst:48]: Good catch :slightly_smiling_face: Fixed.

**Comment by FlashDaggerX** [comment]
Can we add a short section on the appropriate usage of switch cases? 

I'm bringing it up because of its inconsistent use throughout the code base, specifically on `_notification` functions. Some use switches, others use really long `if` / `if-else` blocks.

Examples:
* Neat `switch` cases [(scene/gui/line_edit.cpp)][1]
* Neat `if` block (one case; no `switch` cases) [(scene/gui/panel.cpp)][2]
* Inappropriate use of `if` blocks [(scene/gui/base_button.cpp)][3]
* Inappropriate use of `switch` cases [(scene/2d/line_2d.cpp)][4]
* Another good example of `switch` cases [(scene/main/node.cpp)][5]
* Another example of bad `if` blocks [(scene/main/http_request.cpp)][6]

[1]: <https://github.com/godotengine/godot/blob/d341181f30b4f2e154748911f5d8a59b43aa6e62/scene/gui/line_edit.cpp#L651>
[2]: <https://github.com/godotengine/godot/blob/d341181f30b4f2e154748911f5d8a59b43aa6e62/scene/gui/panel.cpp#L35>
[3]: <https://github.com/godotengine/godot/blob/d341181f30b4f2e154748911f5d8a59b43aa6e62/scene/gui/base_button.cpp#L81>
[4]: <https://github.com/godotengine/godot/blob/d341181f30b4f2e154748911f5d8a59b43aa6e62/scene/2d/line_2d.cpp#L237>
[5]: <https://github.com/godotengine/godot/blob/d341181f30b4f2e154748911f5d8a59b43aa6e62/scene/main/node.cpp#L53>
[6]: <https://github.com/godotengine/godot/blob/d341181f30b4f2e154748911f5d8a59b43aa6e62/scene/main/http_request.cpp#L390>

**Comment by Calinou** [comment]
@FlashDaggerX Thinking about it, I think we should always use `switch` in `_notification` functions, even if there are only 1 or 2 cases. This is to make refactoring easier; when you add new cases, you don't want to spend time converting the `if` block to a `switch` block.

**Comment by Ansraer** [comment]
Could someone explain the reasoning behind disallowing for-range-loops to me?
The docs only mention that they are not explicit enough, but I don't really feel like that's the case. As long as we are only iterating over not overly complex data it should be quite obvious what is happening.

Plus I would argue that the readability and ease of use make them beneficial for "beginner contributors (who may not be
  professional C++ programmers)". In my experience, it is a LOT easier to understand `for(CustomObject obj : list)` than a 70+ characters long loop statement.

**Comment by NathanLovato** [comment]
What's the PR's status? What's missing? Is there a way we can review and merge what's available and consider extending it later?

**Comment by Calinou** [comment]
@NathanLovato I think this is good to merge as-is, but it would be good to get @akien-mga's approval (especially the part about lambdas).

File: community/contributing/cpp_usage_guidelines.rst
Comment (aaronfranke) [inline community/contributing/cpp_usage_guidelines.rst:1]:
```suggestion
.. _doc_cpp_usage_guidelines:
```

**Review by aaronfranke — CHANGES_REQUESTED** [review]

File: community/contributing/cpp_usage_guidelines.rst
Comment (aaronfranke) [inline community/contributing/cpp_usage_guidelines.rst:49]: This causes an error when compiling the docs, not sure what it should be changed to.

File: about/faq.rst
Comment (akien-mga) [inline about/faq.rst:383]: If it's **DoD**, then it should be capitalized **D**ata-**o**riented **D**esign.

File: community/contributing/cpp_usage_guidelines.rst
Comment (akien-mga) [inline community/contributing/cpp_usage_guidelines.rst:19]: While both are true, I don't think they're the only reasons why we use a subset and not full-blown over-engineered C++ (hint in the qualifier). This should likely be discussed further with core devs to come with a strong rationale, otherwise we'll get criticism for taking arbitrary technical decisions just for the sake of non-IDE reviews and junior devs.

File: community/contributing/cpp_usage_guidelines.rst
Comment (akien-mga) [inline community/contributing/cpp_usage_guidelines.rst:27]:
```suggestion
    Prior to Godot 4.0, the C++ standard used throughout the codebase was C++03,
    with a handful of C++11 extensions.
```

File: community/contributing/cpp_usage_guidelines.rst
Comment (akien-mga) [reply community/contributing/cpp_usage_guidelines.rst:49]: @Calinou forgot to add/commit the named reference in the FAQ page.

File: community/contributing/cpp_usage_guidelines.rst
Comment (akien-mga) [inline community/contributing/cpp_usage_guidelines.rst:45]: For once, this can be capitalized in title case as it's a proper name: https://en.wikipedia.org/wiki/Standard_Template_Library

I'd suggest adding a link to that Wikipedia page in the first sentence too to clarify for everyone what we refer to with STL. Up until recently I thought that STL == `std`, while it's specifically about the template components. (So `std::mutex` or `std::thread` are not part of it, and allowed in Godot (core) code which wraps around them).

File: community/contributing/cpp_usage_guidelines.rst
Comment (akien-mga) [inline community/contributing/cpp_usage_guidelines.rst:54]: Maybe we should list all Godot templates and their use case (might need help from some `core` gurus) so that contributors are aware of them. Many use the STL because they're not aware that Godot has its own alternative implementations.

File: community/contributing/cpp_usage_guidelines.rst
Comment (akien-mga) [inline community/contributing/cpp_usage_guidelines.rst:59]: Could use inline code in headings too (I guess you did not because it might not look great, but we do use it already in some places).

File: community/contributing/cpp_usage_guidelines.rst
Comment (akien-mga) [reply community/contributing/cpp_usage_guidelines.rst:89]: I'd say:
```
Lambdas should be used conservatively when they make code effectively faster or simpler, and do not impede readability.
```

File: community/contributing/cpp_usage_guidelines.rst
Comment (akien-mga) [inline community/contributing/cpp_usage_guidelines.rst:96]: Add a seealso or link to https://docs.godotengine.org/en/latest/community/contributing/code_style_guidelines.html#header-includes

File: community/contributing/index.rst
Comment (akien-mga) [inline community/contributing/index.rst:39]: If this is here, `code_style_guidelines` should be next to it IMO.

File: community/contributing/cpp_usage_guidelines.rst
Comment (HaSa1002) [reply community/contributing/cpp_usage_guidelines.rst:54]: Please. I learned C++ as one unit with the standard library and was a afraid of the non-usage of the stuff I learned are standard.

File: community/contributing/cpp_usage_guidelines.rst
Comment (YeldhamDev) [reply community/contributing/cpp_usage_guidelines.rst:19]: Also, the smiley should probably be dropped. It feels kinda condescending.

**Review by fire — APPROVED** [review]
I did a read, and the significant points about autos, STL, and lambdas have been commonly agreed upon by the contributor team.

File: community/contributing/cpp_usage_guidelines.rst
Comment (aaronfranke) [inline community/contributing/cpp_usage_guidelines.rst:75]:
```suggestion
Keep in mind hover documentation isn't often readily available for pull request
```

**Review by aaronfranke — CHANGES_REQUESTED** [review]

**Review by aaronfranke — APPROVED** [review]

**Review by mhilbrunner — APPROVED** [review]

**Comment by Calinou** [comment]
Should this page be pushed to `3.4` as-is, with a note saying that it only applies to the `master` branch? This would help people find it more easily.

**Comment by pycbouh** [comment]
I think that notable differences and general code compatibility warning should be a part of the page anyway. People still support 3.x actively, so we should mention the caveats in the docs. Then it wouldn't matter in which version we store the page, as it would be applicable to the development in general.

**Pull request merged**
